Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m

```diff
@@ -413,83 +315,187 @@ - (void)loadForDisplay
-            [self loadForStillImageDisplay];
+            bodyMediaView = [self loadViewForStillImage];
             break;
         case OWSMessageCellType_AnimatedImage:
             OWSAssert(self.viewItem.attachmentStream);
-            [self loadForAnimatedImageDisplay];
-            break;
-        case OWSMessageCellType_Audio:
-            OWSAssert(self.viewItem.attachmentStream);
-            [self loadForAudioDisplay];
+            bodyMediaView = [self loadViewForAnimatedImage];
             break;
         case OWSMessageCellType_Video:
             OWSAssert(self.viewItem.attachmentStream);
-            [self loadForVideoDisplay];
+            bodyMediaView = [self loadViewForVideo];
             break;
-        case OWSMessageCellType_GenericAttachment: {
+        case OWSMessageCellType_Audio:
             OWSAssert(self.viewItem.attachmentStream);
-            self.attachmentView =
-                [[OWSGenericAttachmentView alloc] initWithAttachment:self.attachmentStream isIncoming:self.isIncoming];
-            [self.attachmentView createContents];
-            [self setMediaView:self.attachmentView];
-            [self addAttachmentUploadViewIfNecessary:self.attachmentView];
-            [self addCaptionIfNecessary];
+            bodyMediaView = [self loadViewForAudio];
+            bodyMediaViewHasGreedyWidth = YES;
             break;
-        }
-        case OWSMessageCellType_DownloadingAttachment: {
-            [self loadForDownloadingAttachment];
-            [self addCaptionIfNecessary];
+        case OWSMessageCellType_GenericAttachment:
+            bodyMediaView = [self loadViewForGenericAttachment];
+            bodyMediaViewHasGreedyWidth = YES;
+            break;
+        case OWSMessageCellType_DownloadingAttachment:
+            bodyMediaView = [self loadViewForDownloadingAttachment];
+            bodyMediaViewHasGreedyWidth = YES;
             break;
+    }
+
+    if (bodyMediaView) {
+        OWSAssert(self.loadCellContentBlock);
+        OWSAssert(self.unloadCellContentBlock);
+        OWSAssert(!lastSubview);
+
+        bodyMediaView.clipsToBounds = YES;
+
+        self.lastBodyMediaView = bodyMediaView;
+        bodyMediaView.userInteractionEnabled = NO;
+        if (self.isMediaBeingSent) {
+            bodyMediaView.layer.opacity = 0.75f;
+        }
+
+        [self.bubbleView addSubview:bodyMediaView];
+        // This layout can lead to extreme cropping of media content,
+        // e.g. a very tall portrait image + long caption.  The media
+        // view will have "max width", so the image will be cropped to
+        // roughly a square.
+        // TODO: Myles is considering alternatives.
+        [self.viewConstraints addObjectsFromArray:@[
+            [bodyMediaView autoPinLeadingToSuperviewWithMargin:0],
+            [bodyMediaView autoPinTrailingToSuperviewWithMargin:0],
+        ]];
+        [NSLayoutConstraint
+            autoSetPriority:UILayoutPriorityDefaultLow
```

Comment:
We need constraints to control the vertical sizing of media and text views, but we use lower priority so that when a message only contains media it uses the exact bounds of the message view.

Reply:
Done.